jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java

```diff
@@ -0,0 +1,506 @@
+                continue;
+            }
+
+            V successor = resultPathVertices.get(deviationIndex + 1);
+            maskedEdges.add(graph.getEdge(pathDeviation, successor));
+        }
+        return Pair.of(maskedVertices, maskedEdges);
+    }
+
+    /**
+     * Builds a candidate path based on the information provided in the
+     * methods parameters. First adds the root part of the candidate by
+     * traversing the vertices and edges of the {@code path} until the
+     * {@code recoverVertexIndex}. Then adds vertices and edges of the
+     * {@code spurPath}.
+     *
+     * @param path               path the candidate path deviates from
+     * @param recoverVertexIndex vertex that is being recovered
+     * @param spurPath           spur path of the candidate
+     * @return candidate path
+     */
+    private GraphPath<V, E> getCandidatePath(GraphPath<V, E> path, int recoverVertexIndex, GraphPath<V, E> spurPath) {
+        List<V> pathVertices = path.getVertexList();
+
+        LinkedList<V> candidatePathVertices = new LinkedList<>();
+        LinkedList<E> candidatePathEdges = new LinkedList<>();
+
+        double rootPathWeight = 0.0;
+        for (int j = 0; j < recoverVertexIndex; j++) {
+            E edge = graph.getEdge(pathVertices.get(j), pathVertices.get(j + 1));
+            rootPathWeight += graph.getEdgeWeight(edge);
+            candidatePathEdges.add(edge);
+            candidatePathVertices.add(pathVertices.get(j));
+        }
+
+        ListIterator<V> spurPathVerticesIterator = spurPath.getVertexList().listIterator(spurPath.getVertexList().size());
+        while (spurPathVerticesIterator.hasPrevious()) {
+            candidatePathVertices.add(spurPathVerticesIterator.previous());
+        }
+        ListIterator<E> spurPathEdgesIterator = spurPath.getEdgeList().listIterator(spurPath.getEdgeList().size());
+        while (spurPathEdgesIterator.hasPrevious()) {
+            candidatePathEdges.add(spurPathEdgesIterator.previous());
+        }
+
+        double candidateWeight = rootPathWeight + spurPath.getWeight();
+        return new GraphWalk<>(graph, source, sink,
+                candidatePathVertices, candidatePathEdges, candidateWeight);
+    }
+
+    /**
+     * Checks if the lists have the same content until the {@code index} (inclusive).
+     *
+     * @param first  first list
+     * @param second second list
+     * @param index  position in the lists
+     * @return true iff the contents of the list are equal until the index
+     */
+    private boolean equalLists(List<V> first, List<V> second, int index) {
+        for (int i = 0; i <= index; i++) {
+            if (!first.get(i).equals(second.get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Helper class which represents the shortest paths tree based on which
+     * the spur paths for the candidates are computed.
```

Comment:
tree using which the spur parts of the paths are computed and appended to the candidate paths